distributed/shuffle/_worker_plugin.py

```diff
@@ -860,10 +869,16 @@ async def teardown(self, worker: Worker) -> None:
         assert not self.closed
 
         self.closed = True
+
         while self.shuffles:
             _, shuffle = self.shuffles.popitem()
-            await shuffle.close()
-            self._runs.remove(shuffle)
+            self.worker._ongoing_background_tasks.call_soon(
+                self._close_shuffle_run, shuffle
+            )
+
+        while self._runs:
+            await asyncio.sleep(0.1)
```

Comment:
I don't insist on this but I don't like these sleep patterns.

```python
def __init__(...):
    self._runs_condition = asyncio.Condition()

async def _close_shuffle_run(self, shuffle: ShuffleRun) -> None:
    await shuffle.close()
    async with self._runs_condition:
        self._runs.remove(shuffle)
        self._runs_condition.notify_all()

async def teardown(self, worker: Worker) -> None:
    ...
    async with self._runs_condition:
        await self._runs_condition.wait_for(lambda: not self._runs)

```

would be a clean alternative. Many people consider Conditions too complex but what I like about them is that they make this relationship very clear (and they unblock **immediately** which is nice for testing and such things).

As I said, I don't insist on this

Reply:
To a more serious question: Is it possible for _runs to be repopulated at this point or are we locking everything up properly for this to not happen once we reach this point?

Reply:
`_runs` is added to in `_refresh_shuffle` which doesn't have a lock associated with it. But I am not sure if that can be running simultaneously with `teardown`.

Reply:
`teardown` sets https://github.com/dask/distributed/blob/90eb9ea1d1991e2d1871dcbbe81f007dfbc8d821/distributed/shuffle/_worker_plugin.py#L890-L893

Once that is done, https://github.com/dask/distributed/blob/90eb9ea1d1991e2d1871dcbbe81f007dfbc8d821/distributed/shuffle/_worker_plugin.py#L810-L811 will raise. We don't yield the event loop in-between that raise and https://github.com/dask/distributed/blob/90eb9ea1d1991e2d1871dcbbe81f007dfbc8d821/distributed/shuffle/_worker_plugin.py#L832